Add support for the `is_multiple` field attribute

sirosen opened the pull request:
This resolves #563

Detection of fields as `is_multiple` for multi-dict unpacking now allows custom fields to use an attribute `is_multiple=T/F` to control the behavior directly. This allows for a subclass of `List` to set `is_multiple=False` (done by `DelimitedList`) and allows non-`List` subclasses to be detected as `multiple`.

Because `is_multiple` is an attribute, it can be set on the field class or on an instance.

A noted TODO comment questions whether or not `ma.fields.Tuple` should be added to the auto-detected fields for `is_multiple`. I wasn't certain if we should add it or not, but I think so? I'll open a follow-up issue to decide on it for 8.0 -- I don't think we should do it in 7.x since 8.0 is on the horizon.

Tests cover
- a field with is_multiple=True
- a field with is_multiple=False
- a field where is_multiple is not set
- a List(Str()) field

and run against the collection of multidict implementations used to test MultiDictProxy.

sirosen committed:
Add support for the `is_multiple` field attribute

This resolves #563

Detection of fields as `is_multiple` for multi-dict unpacking now
allows custom fields to use an attribute `is_multiple=T/F` to control
the behavior directly. This allows for a subclass of `List` to set
`is_multiple=False` (done by `DelimitedList`) and allows non-`List`
subclasses to be detected as `multiple`.

Because `is_multiple` is an attribute, it can be set on the field
class or on an instance.

A noted TODO comment questions whether or not `ma.fields.Tuple` should
be added to the auto-detected fields for `is_multiple`.

Tests cover
- a field with is_multiple=True
- a field with is_multiple=False
- a field where is_multiple is not set
- a List(Str()) field

and run against the collection of multidict implementations used to
test MultiDictProxy.

lafrech commented on src/webargs/core.py:
```diff
@@ -34,6 +32,12 @@
+# which come from marshmallow and therefore don't know about webargs (and
+# do not set `is_multiple=True`)
+# TODO: should `ma.fields.Tuple` be added here?
+KNOWN_MULTI_FIELDS: typing.Tuple[typing.Type, ...] = (ma.fields.List,)
```
Should this be mutable to allow users to add field classes here?

sirosen replied:
I didn't really think about this var as part of the public interface. I'm not against it, I just didn't think of that at all! 😄

If we start going down the path of it being customizable I might want to move it onto the parser so that it can be controlled by subclassing. Right now pretty much all configurable behavior can be set on a parser. Should I take a look at doing that?

lafrech replied:
Sure. Good point.

But no big deal if we keep it as a global. It's not like someone will be instantiating several parsers with different `KNOWN_MULTi_FIELD` values.

sirosen replied:
I think you're right that multiple parsers with different values is an unlikely/strange use-case.

It's getting late in my local time (US Eastern), so I'll come back to this tomorrow. But right now I'm inclined to change it to be a list and merge the result. At least that way, it's not actively stopping anyone from modifying it.

lafrech replied:
Yeah, it is not a really friendly API, but it doesn't get in the way of advanced users.

lafrech reviewed (approved):
Sorry for the delay.

LGTM.

I'd be tempted to consider #585 a bugfix, but if 8.0 is on its way, this can wait.

The feature might deserve to be documented but this is not a blocker. We can release this as is in 7.x for curious users and advertise it in the docs later.

sirosen commented:
> I'd be tempted to consider #585 a bugfix, but if 8.0 is on its way, this can wait.

Yeah, if we were going to take a long time to do 8.0, I would maybe be more inclined to put it in as a bugfix; maybe included here. But I just want to take care of whatever can be done without a major version bump and then go ahead and do 8.0 with the changes we've planned.

I could add something in the advanced doc on this... Let's circle back on that later though? I think the changelog is doc enough for the moment.

sirosen committed:
Convert KNOWN_MULTI_FIELDS to a list

By using a list rather than a tuple, we make it easier for advanced
users to monkeypatch and adjust the behavior of webargs. This is still
an implementation detail (not a public interface).

sirosen commented:
I converted `KNOWN_MULTI_FIELDS` to a list ( 7f93baa ) and merged on the basis of the prior approval.
I'll take a look at making that configurable and documented as a parser attribute.
